### PR TITLE
fixed error in Command sonata:admin:explain for symfony 2.* and  symfony 3.* 

### PR DESCRIPTION
**Pull request by rispo-service**
### Fixed

Error for sonata:admin:explain Command in Symfony2.x and 3.x version.

### CODE REVIEW

**Comment by greg0ire** [comment]
Hi, and thanks for your contribution. Can you show us the error you are getting?
Please associate your email address with your github account.

File: Command/ExplainAdminCommand.php [hunk 1]
```diff
@@ -89,11 +89,7 @@ public function execute(InputInterface $input, OutputInterface $output)
             $output->writeln(sprintf('  - % -25s  % -15s % -15s', $name, $fieldDescription->getType(), $fieldDescription->getTemplate()));
         }
 
-        if ($this->getContainer()->has('validator.validator_factory')) {
-            $metadata = $this->getContainer()->get('validator.validator_factory')->getMetadataFor($admin->getClass());
-        } else {
-            $metadata = $this->getContainer()->get('validator')->getMetadataFor($admin->getClass());
-        }
+        $metadata = $this->getContainer()->get('validator')->getMetadataFor($admin->getClass());
```
Comment (greg0ire) [inline Command/ExplainAdminCommand.php]: Can explain how this fixes the error?

Comment (rispo-service) [inline Command/ExplainAdminCommand.php]: Error:

#: php app/console sonata:admin:list
....cut...

[Symfony\Component\Debug\Exception\UndefinedMethodException]
  Attempted to call an undefined method named "getMetadataFor" of class "Symfony\Bundle\FrameworkBundle\Validator\ConstraintValidatorFactory".

Really: [ConstraintValidatorFactory.php](https://github.com/symfony/framework-bundle/blob/master/Validator/ConstraintValidatorFactory.php)

**Comment by OskarStark** [comment]
When i execute:
```
app/console sonata:admin:explain channel.admin.channel
```

I get the following result:
```
AdminBundle Information
id                   : channel.admin.channel
Admin                : ChannelBundle\Admin\ChannelAdmin
Model                : ChannelBundle\Entity\Channel
Controller           : SonataAdminBundle:CRUD
Model Manager        : Sonata\DoctrineORMAdminBundle\Model\ModelManager
Form Builder         : Symfony\Component\Form\FormBuilder
Datagrid Builder     : Sonata\DoctrineORMAdminBundle\Builder\DatagridBuilder
List Builder         : Sonata\DoctrineORMAdminBundle\Builder\ListBuilder

Routes
  - admin_channel_channel_list /channel/channel/list
  - admin_channel_channel_create /channel/channel/create
  - admin_channel_channel_batch /channel/channel/batch
  - admin_channel_channel_edit /channel/channel/{id}/edit
  - admin_channel_channel_delete /channel/channel/{id}/delete
  - admin_channel_channel_show /channel/channel/{id}/show
  - admin_channel_channel_export /channel/channel/export
  - admin_channel_channel_history /channel/channel/{id}/history
  - admin_channel_channel_history_view_revision /channel/channel/{id}/history/{revision}/view
  - admin_channel_channel_history_compare_revisions /channel/channel/{id}/history/{base_revision}/{compare_revision}/compare

Datagrid Columns
  - name                       text            SonataAdminBundle:CRUD:base_list_field.html.twig
  - parent                     orm_many_to_one SonataDoctrineORMAdminBundle:CRUD:list_orm_many_to_one.html.twig
  - teleplanMappingName        text            SonataAdminBundle:CRUD:base_list_field.html.twig
  - isMainChannel              boolean         SonataAdminBundle:CRUD:list_boolean.html.twig
  - reel                       text            SonataAdminBundle:CRUD:base_list_field.html.twig
  - looprecorders              orm_many_to_many SonataDoctrineORMAdminBundle:CRUD:list_orm_many_to_many.html.twig
  - _action                    actions         SonataAdminBundle:CRUD:list__action.html.twig

Datagrid Filters
  - name                       doctrine_orm_string
  - teleplanMappingName        doctrine_orm_string
  - parent                     doctrine_orm_model
  - isMainChannel              doctrine_orm_boolean
  - datasetName                doctrine_orm_string
  - oegs                       doctrine_orm_boolean
  - audioDescription           doctrine_orm_boolean
  - bitmovinStreamId           doctrine_orm_string
  - createdAt                  doctrine_orm_datetime
  - updatedAt                  doctrine_orm_datetime

Form theme(s)
  - SonataDoctrineORMAdminBundle:Form:form_admin_fields.html.twig
  - ClipEditorBundle:Form:fields.html.twig

Form Fields
  - name                       string
  - parent                     sonata_type_model_list
  - teleplanMappingName        string
  - isMainChannel              boolean
  - frontendCssClass           string
  - reel                       choice
  - looprecorders              sonata_type_model
  - datasetName                string
  - oegs                       boolean
  - audioDescription           boolean
  - bitmovinStreamId           string
  - colorLogo                  sonata_media_type
  - blackAndWhiteLogo          sonata_media_type
  - relatedOegsChannel         sonata_type_model_list
  - relatedAudioDescriptionChannel  sonata_type_model_list


  [Symfony\Component\Debug\Exception\UndefinedMethodException]
  Attempted to call an undefined method named "getMetadataFor" of class "Symfony\Bundle\FrameworkBun
  dle\Validator\ConstraintValidatorFactory".


sonata:admin:explain [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-s|--shell] [--process-isolation] [-e|--env ENV] [--no-debug] [--] <command> <admin>
```

The error:
```
  [Symfony\Component\Debug\Exception\UndefinedMethodException]
  Attempted to call an undefined method named "getMetadataFor" of class "Symfony\Bundle\FrameworkBun
  dle\Validator\ConstraintValidatorFactory".
```

Used versions:
```
ass/xmlsecurity                          v1.1.1  The XmlSecurity library is written in PHP for wor...
behat/behat                              v3.1.0  Scenario-oriented BDD framework for PHP 5.3
behat/gherkin                            v4.4.1  Gherkin DSL parser for PHP 5.3
behat/mink                               v1.7.1  Browser controller/emulator abstraction for PHP
behat/mink-browserkit-driver             v1.3.2  Symfony2 BrowserKit driver for Mink framework
behat/mink-extension                     v2.2    Mink extension for Behat
behat/mink-selenium2-driver              v1.3.1  Selenium2 (WebDriver) driver for Mink framework
behat/symfony2-extension                 2.1.1   Symfony2 framework extension for Behat
behat/transliterator                     v1.1.0  String transliterator
besimple/soap                            v0.2.6  Build and consume SOAP and WSDL based web services
c33s/symfony-config-manipulator-bundle   v0.2.2  Utilities to automatically organize Symfony2 YAML...
cocur/slugify                            v2.2    Converts a string into a slug.
composer/installers                      v1.1.0  A multi-framework Composer library installer
doctrine/annotations                     v1.2.7  Docblock Annotations Parser
doctrine/cache                           v1.6.0  Caching library offering an object-oriented API f...
doctrine/collections                     v1.3.0  Collections Abstraction library
doctrine/common                          v2.6.1  Common Library for Doctrine projects
doctrine/data-fixtures                   v1.2.1  Data Fixtures for all Doctrine Object Managers
doctrine/dbal                            v2.5.4  Database Abstraction Layer
doctrine/doctrine-bundle                 1.6.3   Symfony DoctrineBundle
doctrine/doctrine-cache-bundle           1.3.0   Symfony Bundle for Doctrine Cache
doctrine/doctrine-fixtures-bundle        2.3.0   Symfony DoctrineFixturesBundle
doctrine/doctrine-migrations-bundle      1.1.1   Symfony DoctrineMigrationsBundle
doctrine/inflector                       v1.1.0  Common String Manipulations with regard to casing...
doctrine/instantiator                    1.0.5   A small, lightweight utility to instantiate objec...
doctrine/lexer                           v1.0.1  Base library for a lexer that can be used in Top-...
doctrine/migrations                      1.4.1   Database Schema migrations using Doctrine DBAL
doctrine/orm                             v2.5.4  Object-Relational-Mapper for PHP
egeloen/ckeditor-bundle                  4.0.4   Provides a CKEditor integration for your Symfony2...
egeloen/json-builder                     2.0.1   JSON builder with escaping control for PHP 5.3+
friendsofphp/php-cs-fixer                v1.11.6 A tool to automatically fix PHP code style
friendsofsymfony/elastica-bundle         3.1.8   Elasticsearch PHP integration for your Symfony2 p...
friendsofsymfony/http-cache              1.4.2   Tools to manage cache invalidation
friendsofsymfony/http-cache-bundle       1.3.7   Set path based HTTP cache headers and send invali...
friendsofsymfony/jsrouting-bundle        1.6.0   A pretty nice way to expose your Symfony2 routing...
friendsofsymfony/rest-bundle             1.8.0   This Bundle provides various tools to rapidly dev...
friendsofsymfony/user-bundle             v1.3.6  Symfony FOSUserBundle
gedmo/doctrine-extensions                v2.4.20 Doctrine2 behavioral extensions
guzzle/guzzle                            v3.9.3  PHP HTTP client. This library is deprecated in fa...
imagine/imagine                          v0.6.3  Image processing for PHP 5.3
incenteev/composer-parameter-handler     v2.1.2  Composer script handling your ignored parameter file
instaclick/php-webdriver                 1.4.3   PHP WebDriver for Selenium 2
ircmaxell/password-compat                v1.0.4  A compatibility library for the proposed simplifi...
jdorn/sql-formatter                      v1.2.17 a PHP SQL highlighting library
jms/metadata                             1.5.1   Class/method/property metadata management in PHP
jms/parser-lib                           1.0.0   A library for easily creating recursive-descent p...
jms/serializer                           1.2.0   Library for (de-)serializing data of any complexi...
jms/serializer-bundle                    1.1.0   Allows you to easily serialize, and deserialize d...
jublonet/codebird-php                    3.1.0   Easy access to the Twitter REST API, Collections ...
knplabs/gaufrette                        v0.1.9  PHP5 library that provides a filesystem abstracti...
knplabs/knp-components                   1.3.3   Knplabs component library
knplabs/knp-markdown-bundle              1.5.1   Knplabs markdown bundle transforms markdown into ...
knplabs/knp-menu                         v2.1.1  An object oriented menu library
knplabs/knp-menu-bundle                  2.1.2   This bundle provides an integration of the KnpMen...
knplabs/knp-paginator-bundle             2.5.3   Paginator bundle for Symfony2 to automate paginat...
kphoen/doctrine-state-machine            1.1.0   Doctrine2 state machine behavior
kphoen/doctrine-state-machine-bundle     1.1.2   Integrates DoctrineStateMachineBehavior in Symfony2.
kriswallsmith/assetic                    v1.3.2  Asset Management for PHP
kriswallsmith/buzz                       v0.15   Lightweight HTTP client
league/flysystem                         1.0.26  Filesystem abstraction: Many filesystems, one API.
league/flysystem-cached-adapter          1.0.3   An adapter decorator to enable meta-data caching.
league/flysystem-sftp                    1.0.10  Flysystem adapter for SFTP
michelf/php-markdown                     1.6.0   PHP Markdown
monolog/monolog                          1.21.0  Sends your logs to files, sockets, inboxes, datab...
myclabs/deep-copy                        1.5.1   Create deep copies (clones) of your objects
nelmio/api-doc-bundle                    2.13.0  Generates documentation for your REST API from an...
nelmio/cors-bundle                       1.4.1   Adds CORS (Cross-Origin Resource Sharing) headers...
ocramius/package-versions                1.1.1   Composer plugin that provides efficient querying ...
ocramius/proxy-manager                   2.0.3   A library providing utilities to generate, instan...
pagerfanta/pagerfanta                    v1.0.3  Pagination for PHP 5.3
paragonie/random_compat                  v2.0.2  PHP 5.x polyfill for random_bytes() and random_in...
patchwork/utf8                           v1.3.1  Portable and performant UTF-8, Unicode and Graphe...
pdepend/pdepend                          2.2.4   Official version of pdepend to be handled with Co...
phing/phing                              2.14.0  PHing Is Not GNU make; it's a PHP project build s...
phpcollection/phpcollection              0.4.0   General-Purpose Collection Library for PHP
phpdocumentor/reflection-common          1.0     Common reflection classes used by phpdocumentor t...
phpdocumentor/reflection-docblock        3.1.0   With this component, a library can provide suppor...
phpdocumentor/type-resolver              0.2
phpmd/phpmd                              2.4.3   PHPMD is a spin-off project of PHP Depend and aim...
phpoption/phpoption                      1.5.0   Option Type for PHP
phpseclib/phpseclib                      2.0.2   PHP Secure Communications Library - Pure-PHP impl...
phpspec/prophecy                         v1.6.1  Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage                4.0.1   Library that provides collection, processing, and...
phpunit/php-file-iterator                1.4.1   FilterIterator implementation that filters files ...
phpunit/php-text-template                1.2.1   Simple template engine.
phpunit/php-timer                        1.0.8   Utility class for timing
phpunit/php-token-stream                 1.4.8   Wrapper around PHP's tokenizer extension.
phpunit/phpunit                          5.4.8   The PHP Unit Testing framework.
phpunit/phpunit-mock-objects             3.2.3   Mock Object library for PHPUnit
pixassociates/sortable-behavior-bundle   v1.0    Provides a way to sort your admin listing
propel/propel-bundle                     1.5.1   Integration of Propel in Symfony2
propel/propel1                           1.7.1   Propel is an open-source Object-Relational Mappin...
psr/cache                                1.0.0   Common interface for caching libraries
psr/log                                  1.0.0   Common interface for logging libraries
ramsey/uuid                              3.5.0   Formerly rhumsaa/uuid. A PHP 5.4+ library for gen...
ramsey/uuid-doctrine                     1.2.0   Allow the use of a ramsey/uuid UUID as Doctrine f...
ruflin/elastica                          2.3.1   Elasticsearch Client
salavert/time-ago-in-words               v1.6.2  Time ago Twig extension where you can easily conv...
sebastian/code-unit-reverse-lookup       1.0.0   Looks up which function or method a line of code ...
sebastian/comparator                     1.2.0   Provides the functionality to compare PHP values ...
sebastian/diff                           1.4.1   Diff implementation
sebastian/environment                    1.3.7   Provides functionality to handle HHVM/PHP environ...
sebastian/exporter                       1.2.2   Provides the functionality to export PHP variable...
sebastian/global-state                   1.1.1   Snapshotting of global state
sebastian/object-enumerator              1.0.0   Traverses array structures and object graphs to e...
sebastian/recursion-context              1.0.2   Provides functionality to recursively process PHP...
sebastian/resource-operations            1.0.0   Provides a list of PHP built-in functions that op...
sebastian/version                        2.0.0   Library that helps with managing the version numb...
sensio/distribution-bundle               v4.0.7  Base bundle for Symfony Distributions
sensio/framework-extra-bundle            v3.0.12 This bundle provides a way to configure your cont...
sensio/generator-bundle                  v2.5.3  This bundle generates code for you
sensiolabs/security-checker              v3.0.2  A security checker for your composer.lock
simplethings/entity-audit-bundle         0.9.1   Audit for Doctrine Entities
sonata-project/admin-bundle              3.4.0   The missing Symfony Admin Generator
sonata-project/block-bundle              3.1.1   Symfony SonataBlockBundle
sonata-project/cache                     1.0.7   Cache library
sonata-project/cache-bundle              2.2.5   This bundle provides caching services
sonata-project/classification-bundle     3.1.0   Symfony SonataClassificationBundle
sonata-project/core-bundle               3.0.3   Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.2     Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2   Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.0.5   Symfony Sonata / Integrate Doctrine ORM into the ...
sonata-project/easy-extends-bundle       2.1.10  Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.6.0   Lightweight Exporter library
sonata-project/formatter-bundle          3.0.1   Symfony SonataFormatterBundle
sonata-project/google-authenticator      1.0.2   Library to integrate Google Authenticator into a ...
sonata-project/intl-bundle               2.2.4   Symfony SonataIntlBundle
sonata-project/media-bundle              3.1.0   Symfony SonataMediaBundle
sonata-project/notification-bundle       3.0.0   Symfony SonataNotificationBundle
sonata-project/page-bundle               3.1.0   This bundle provides a Site and Page management t...
sonata-project/seo-bundle                2.0.2   Symfony SonataSeoBundle
sonata-project/timeline-bundle           3.0.0   Integrates SpyTimelineBundle into Sonata
sonata-project/user-bundle               3.0.1   Symfony SonataUserBundle
spe/filesize-extension-bundle            1.0.0   Twig filter for human readable filesize
squizlabs/php_codesniffer                2.6.2   PHP_CodeSniffer tokenizes PHP, JavaScript and CSS...
stephpy/timeline                         v1.0.4  Standalone library to create timelines
stephpy/timeline-bundle                  2.3.0   Symfony2 bundle to make timeline
stof/doctrine-extensions-bundle          v1.2.2  Integration of the gedmo/doctrine-extensions with...
swiftmailer/swiftmailer                  v5.4.3  Swiftmailer, free feature-rich PHP mailer
symfony-cmf/routing                      1.4.0   Extends the Symfony2 routing component for dynami...
symfony-cmf/routing-bundle               1.4.0   Symfony RoutingBundle
symfony/assetic-bundle                   v2.8.0  Integrates Assetic into Symfony2
symfony/monolog-bundle                   2.11.1  Symfony MonologBundle
symfony/phpunit-bridge                   v2.8.9  Symfony PHPUnit Bridge
symfony/polyfill-apcu                    v1.2.0  Symfony polyfill backporting apcu_* functions to ...
symfony/polyfill-intl-icu                v1.2.0  Symfony polyfill for intl's ICU-related data and ...
symfony/polyfill-mbstring                v1.2.0  Symfony polyfill for the Mbstring extension
symfony/polyfill-php54                   v1.2.0  Symfony polyfill backporting some PHP 5.4+ featur...
symfony/polyfill-php55                   v1.2.0  Symfony polyfill backporting some PHP 5.5+ featur...
symfony/polyfill-php56                   v1.2.0  Symfony polyfill backporting some PHP 5.6+ featur...
symfony/polyfill-php70                   v1.2.0  Symfony polyfill backporting some PHP 7.0+ featur...
symfony/polyfill-util                    v1.2.0  Symfony utilities for portability of PHP codes
symfony/security-acl                     v3.0.0  Symfony Security Component - ACL (Access Control ...
symfony/swiftmailer-bundle               v2.3.11 Symfony SwiftmailerBundle
symfony/symfony                          v2.8.9  The Symfony PHP framework
twig/extensions                          v1.3.0  Common additional features for Twig that do not d...
twig/twig                                v1.24.1 Twig, the flexible, fast, and secure template lan...
webmozart/assert                         1.0.2   Assertions to validate method input/output with n...
willdurand/hateoas                       2.10.0  A PHP library to support implementing representat...
willdurand/hateoas-bundle                1.2.0   Integration of Hateoas into Symfony2.
willdurand/jsonp-callback-validator      v1.1.0  JSONP callback validator.
willdurand/negotiation                   1.5.0   Content Negotiation tools for PHP provided as a s...
willdurand/propel-typehintable-behavior  1.0.5   Insane Propel behavior that helps you to be compl...
yohang/finite                            1.1.0   A simple PHP5.3+ Finite State Machine
zendframework/zend-code                  3.0.4   provides facilities to generate arbitrary code us...
zendframework/zend-eventmanager          3.0.1   Trigger and listen to events within a PHP applica...
zendframework/zend-mime                  2.1.6
zendframework/zend-stdlib                2.1.6
zendframework/zenddiagnostics            v1.0.9  A set of components for performing diagnostic tes..
```

After the change of @rispo-service I get the correct result:
```
AdminBundle Information
id                   : channel.admin.channel
Admin                : ChannelBundle\Admin\ChannelAdmin
Model                : ChannelBundle\Entity\Channel
Controller           : SonataAdminBundle:CRUD
Model Manager        : Sonata\DoctrineORMAdminBundle\Model\ModelManager
Form Builder         : Symfony\Component\Form\FormBuilder
Datagrid Builder     : Sonata\DoctrineORMAdminBundle\Builder\DatagridBuilder
List Builder         : Sonata\DoctrineORMAdminBundle\Builder\ListBuilder

Routes
  - admin_channel_channel_list /channel/channel/list
  - admin_channel_channel_create /channel/channel/create
  - admin_channel_channel_batch /channel/channel/batch
  - admin_channel_channel_edit /channel/channel/{id}/edit
  - admin_channel_channel_delete /channel/channel/{id}/delete
  - admin_channel_channel_show /channel/channel/{id}/show
  - admin_channel_channel_export /channel/channel/export
  - admin_channel_channel_history /channel/channel/{id}/history
  - admin_channel_channel_history_view_revision /channel/channel/{id}/history/{revision}/view
  - admin_channel_channel_history_compare_revisions /channel/channel/{id}/history/{base_revision}/{compare_revision}/compare

Datagrid Columns
  - name                       text            SonataAdminBundle:CRUD:base_list_field.html.twig
  - parent                     orm_many_to_one SonataDoctrineORMAdminBundle:CRUD:list_orm_many_to_one.html.twig
  - teleplanMappingName        text            SonataAdminBundle:CRUD:base_list_field.html.twig
  - isMainChannel              boolean         SonataAdminBundle:CRUD:list_boolean.html.twig
  - reel                       text            SonataAdminBundle:CRUD:base_list_field.html.twig
  - looprecorders              orm_many_to_many SonataDoctrineORMAdminBundle:CRUD:list_orm_many_to_many.html.twig
  - _action                    actions         SonataAdminBundle:CRUD:list__action.html.twig

Datagrid Filters
  - name                       doctrine_orm_string
  - teleplanMappingName        doctrine_orm_string
  - parent                     doctrine_orm_model
  - isMainChannel              doctrine_orm_boolean
  - datasetName                doctrine_orm_string
  - oegs                       doctrine_orm_boolean
  - audioDescription           doctrine_orm_boolean
  - bitmovinStreamId           doctrine_orm_string
  - createdAt                  doctrine_orm_datetime
  - updatedAt                  doctrine_orm_datetime

Form theme(s)
  - SonataDoctrineORMAdminBundle:Form:form_admin_fields.html.twig
  - ClipEditorBundle:Form:fields.html.twig

Form Fields
  - name                       string
  - parent                     sonata_type_model_list
  - teleplanMappingName        string
  - isMainChannel              boolean
  - frontendCssClass           string
  - reel                       choice
  - looprecorders              sonata_type_model
  - datasetName                string
  - oegs                       boolean
  - audioDescription           boolean
  - bitmovinStreamId           string
  - colorLogo                  sonata_media_type
  - blackAndWhiteLogo          sonata_media_type
  - relatedOegsChannel         sonata_type_model_list
  - relatedAudioDescriptionChannel  sonata_type_model_list

Validation Framework - http://symfony.com/doc/3.0/book/validation.html
Properties constraints
    no property constraints defined !!

Getters constraints
    no getter constraints defined !!
```

File: Command/ExplainAdminCommand.php (hunk 1 above)
Comment (greg0ire) [inline Command/ExplainAdminCommand.php]: Oh wow, that's strange.

**Comment by greg0ire** [comment]
Looks like the tests have to be changed too…

**Comment by greg0ire** [comment]
ping @rispo-service , do you need help fixing the tests?

**Comment by OskarStark** [comment]
I close this PR, because the fix is introduced and merged in #4083 

@rispo-service you can try the latest 3.x branch, until this is released and check if this fixes your problem

thank you for your contribution

**Pull request closed without merging**